File: CppSolution/MergeIntervals.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include "Utility.h"
using namespace std;

class Solution
{
public:
    static bool myfunc(const Interval &a, const Interval &b)
    {
        return (a.start < b.start);
    }
    vector<Interval> merge(vector<Interval> &intervals)
    {
        vector<Interval> result;
        if (intervals.size() == 1 || intervals.size() == 0)
            return intervals;
        sort(intervals.begin(), intervals.end(), Solution::myfunc);
        result.push_back(*(intervals.begin()));
        vector<Interval>::iterator it = intervals.begin() + 1;
        while (it != intervals.end())
        {
            if (result[result.size() - 1].end >= (it)->start)
            {
                result[result.size() - 1].end = max((it)->end, result[result.size() - 1].end);                
            }
            else
            {
                result.push_back(*it);
            }
            ++it;
        }
        return intervals;
    }
};

int main()
{
    return 0;
}

File: JavaSolution/src/Google/MergeIntervals.java
package Google;

/**
 * Created by xyunpeng on 2/14/16.
 */
public class MergeIntervals {
    public class IntervalComparator implements Comparator<Interval> {
        public int compare(Interval i1, Interval i2) {
            return i1.start - i2.start;
        }
    }

    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> result = new ArrayList<>();

        if (intervals == null || intervals.size() == 0) {
            return result;
        }

        Collections.sort(intervals, new IntervalComparator());
        Interval last = new Interval(intervals.get(0).start, intervals.get(0).end);
        for (int i = 1; i < intervals.size(); i++) {
            if (last.end < intervals.get(i).start) {
                result.add(last);
                last = new Interval(intervals.get(i).start, intervals.get(i).end);
            } else {
                last.end = Math.max(last.end, intervals.get(i).end);
            }
        }
        result.add(last);
        return result;
    }
}
